### SET MEMBERSHIP

+ Sets seems to be Similar to Lists and Tuples.
+ But, more Restricted.
+ So why we should use Sets...?


+ **Sets can be used to fix Bugs**

In [6]:
## simple menu program

## initializing choice
choice = "-"

# use a while loop for choice not equals to 0,
# whenever choice is 0,
# program will exit the while loop
while choice != "0":
    
    ## implement a if, else statements
    ## that will 
    ## if choice is in the string 12345
    ## print the choice.
    ## else ask to print from the below choices.
    ## and print the all available choices.
    if choice in "12345":
        print("You Chose {}".format(choice))
        
    else:
        print("Please select the value from the Below Choices.")
        print("1 : \tLearn Python")
        print("2 : \tLearn Java")
        print("3 : \tGo Swimming")
        print("4 : \tHave Dinner")
        print("5 : \tGo to Bed")
        print("0 : \tExit")
        
    ## get the choice from here as the input every time the loops works
    choice = input()

Please select the value from the Below Choices.
1 : 	Learn Python
2 : 	Learn Java
3 : 	Go Swimming
4 : 	Have Dinner
5 : 	Go to Bed
0 : 	Exit


 123


You Chose 123


You Chose 


 0


+ Here, if we choose "123", this also will give us an Output.
+ But that does not contain in the Menu.
+ Also, we can give an entry as an empty string.
+ That will also give us an Output which is not in the Menu.
+ So that is a Bug.

In [8]:
## if we check, that the inputis in a list instead of a string, the bugs would go away.
## changing the program
## simple menu program

## initializing choice
choice = "-"

# use a while loop for choice not equals to 0,
# whenever choice is 0,
# program will exit the while loop
while choice != "0":
    
    ## implement a if, else statements
    ## that will 
    ## if choice is in the string 12345
    ## print the choice.
    ## else ask to print from the below choices.
    ## and print the all available choices.       ## list("12345") = ['1', '2', '3', '4', '5']
    if choice in list("12345"):                   ## iterating over the list("12345") list of string 12345
        print("You Chose {}".format(choice))
        
    else:
        print("Please select the value from the Below Choices.")
        print("1 : \tLearn Python")
        print("2 : \tLearn Java")
        print("3 : \tGo Swimming")
        print("4 : \tHave Dinner")
        print("5 : \tGo to Bed")
        print("0 : \tExit")
        
    ## get the choice from here as the input every time the loops works
    choice = input()

Please select the value from the Below Choices.
1 : 	Learn Python
2 : 	Learn Java
3 : 	Go Swimming
4 : 	Have Dinner
5 : 	Go to Bed
0 : 	Exit


 12


Please select the value from the Below Choices.
1 : 	Learn Python
2 : 	Learn Java
3 : 	Go Swimming
4 : 	Have Dinner
5 : 	Go to Bed
0 : 	Exit


 23


Please select the value from the Below Choices.
1 : 	Learn Python
2 : 	Learn Java
3 : 	Go Swimming
4 : 	Have Dinner
5 : 	Go to Bed
0 : 	Exit


 34


Please select the value from the Below Choices.
1 : 	Learn Python
2 : 	Learn Java
3 : 	Go Swimming
4 : 	Have Dinner
5 : 	Go to Bed
0 : 	Exit


 0


+ Here list("12345") = ['1', '2', '3', '4', '5']
+ So if we choose sequnces like "123"
+ We will get the menu printed as earlierly Expected.
+ Here using the List is good.
+ But more appropriate one is Set.


In [11]:
## if we check, that the inputis in a set instead of a string, the bugs would go away.
## changing the program
## simple menu program

## initializing choice
choice = "-"

# use a while loop for choice not equals to 0,
# whenever choice is 0,
# program will exit the while loop
while choice != "0":
    
    ## implement a if, else statements
    ## that will 
    ## if choice is in the string 12345
    ## print the choice.
    ## else ask to print from the below choices.
    ## and print the all available choices.       ## set("12345") = {'1', '2', '3', '4', '5'}
    if choice in set("12345"):                   ## iterating over the set("12345") list of string 12345
        print("You Chose {}".format(choice))
        
    else:
        print("Please select the value from the Below Choices.")
        print("1 : \tLearn Python")
        print("2 : \tLearn Java")
        print("3 : \tGo Swimming")
        print("4 : \tHave Dinner")
        print("5 : \tGo to Bed")
        print("0 : \tExit")
        
    ## get the choice from here as the input every time the loops works
    choice = input()

Please select the value from the Below Choices.
1 : 	Learn Python
2 : 	Learn Java
3 : 	Go Swimming
4 : 	Have Dinner
5 : 	Go to Bed
0 : 	Exit


 1


You Chose 1


 2


You Chose 2


 23


Please select the value from the Below Choices.
1 : 	Learn Python
2 : 	Learn Java
3 : 	Go Swimming
4 : 	Have Dinner
5 : 	Go to Bed
0 : 	Exit


Please select the value from the Below Choices.
1 : 	Learn Python
2 : 	Learn Java
3 : 	Go Swimming
4 : 	Have Dinner
5 : 	Go to Bed
0 : 	Exit


 345


Please select the value from the Below Choices.
1 : 	Learn Python
2 : 	Learn Java
3 : 	Go Swimming
4 : 	Have Dinner
5 : 	Go to Bed
0 : 	Exit


 0


In [16]:
## list function
list_function = list("12345")
print(list_function)

set_function = set("12345")
print(set_function)


## we can use range function as well along with set.
range_function_with_Set = set(range(0,6,3))
print(range_function_with_Set)

['1', '2', '3', '4', '5']
{'1', '4', '5', '3', '2'}
{0, 3}


### Testing set membership is Fast.

+ If you are working with larger data, checking for membership is lot faster with a set compared to a List.
+ 